core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -412,12 +416,69 @@ public final SegmentsStats segmentsStats() {
             SegmentsStats stats = new SegmentsStats();
             for (LeafReaderContext reader : searcher.reader().leaves()) {
                 final SegmentReader segmentReader = segmentReader(reader.reader());
+                SegmentsStats.IndexResources memoryResources = new SegmentsStats.IndexResources();
+                SegmentsStats.IndexResources diskResources = new SegmentsStats.IndexResources();
                 stats.add(1, segmentReader.ramBytesUsed());
-                stats.addTermsMemoryInBytes(guardedRamBytesUsed(segmentReader.getPostingsReader()));
-                stats.addStoredFieldsMemoryInBytes(guardedRamBytesUsed(segmentReader.getFieldsReader()));
-                stats.addTermVectorsMemoryInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
-                stats.addNormsMemoryInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
-                stats.addDocValuesMemoryInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+                memoryResources.addTermsInBytes(guardedRamBytesUsed(segmentReader.getPostingsReader()));
+                memoryResources.addStoredFieldsInBytes(guardedRamBytesUsed(segmentReader.getFieldsReader()));
+                memoryResources.addTermVectorsInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
+                memoryResources.addNormsInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
+                memoryResources.addDocValuesInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+                stats.addMemoryIndexResources(memoryResources);
+                
+                Directory directory = null;
+                SegmentInfo segmentInfo = segmentReader.getSegmentInfo().info;
+                boolean useCompoundFile = segmentInfo.getUseCompoundFile();
+                if (useCompoundFile) {
+                    try {
+                        directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentInfo, IOContext.READ);
+                    } catch (IOException e) {
+                        logger.trace("error when opening compound reader for segment [{}]", e, segmentInfo.name);
+                    }
+                } else {
+                    directory = segmentReader.directory();
+                }
+                
+                // Couldn't open compound reader, skip
+                if (directory == null) {
+                    continue;
+                }
+                
+                try {
+                    String[] files = directory.listAll();
```

Comment:
`listAll` and pulling `fileLength` can be costly operations (much more so than pulling memory stats) especially when IO is stressed  ... I wonder if we should make this (pull disk stats in addition to memory stats) an option on the stats request?